packages/@atjson/peritext/src/index.ts

```diff
@@ -0,0 +1,585 @@
+import Document, {
+  Block,
+  Mark,
+  EdgeBehaviour,
+  JSON,
+  AnnotationConstructor,
+  AttributesOf,
+  Annotation,
+  SliceAnnotation,
+  TextAnnotation,
+} from "@atjson/document";
+import uuid from "uuid-random";
+
+export type Peritext = {
+  text: string;
+  blocks: Block[];
+  marks: Mark[];
+};
+
+type Range = `${"[" | "("}${number}..${number}${"]" | ")"}`;
+
+function areParentsEqual(left: string[], right: string[]) {
+  return (
+    left.length === right.length &&
+    left.every((value, index) => value === right[index])
+  );
+}
+
+function attributesWithStableIds(
+  attribute: any,
+  ids: Map<string, string>
+): NonNullable<any> {
+  if (attribute == null) {
+    return null;
+  } else if (Array.isArray(attribute)) {
+    const copy: any[] = [];
+    for (let i = 0, len = attribute.length; i < len; i++) {
+      copy[i] = attributesWithStableIds(attribute[i], ids);
+    }
+    return copy;
+  } else if (attribute instanceof Document) {
+    return attribute;
+  } else if (typeof attribute === "object") {
+    const copy: NonNullable<any> = {};
+    for (const key in attribute) {
+      if (attribute[key] !== undefined) {
+        copy[key] = attributesWithStableIds(attribute[key], ids);
+      }
+    }
+    return copy;
+  } else if (typeof attribute === "string" && ids.has(attribute)) {
+    return ids.get(attribute);
+  }
+  return attribute;
+}
+
+/**
+ * Returns a modified copy of `doc` where all blocks and
+ *   marks have had their `id` properties re-generated in a
+ *   deterministic way, to enable easy document comparisons during
+ *   testing.
+ *
+ * Any properties on blocks or marks that exactly match a block or
+ *   mark `id` property will also be updated to the corresponding
+ *   id in the resulting document. Only exact matches are treated
+ *   this way, substring matches are not detected or handled.
+ *
+ * @param doc - a peritext document
+ * @returns a modified copy of `doc` with stabilized ids
+ */
+export function stabilizeIds(doc: Peritext) {
+  let blockCounter = 0;
+  let markCounter = 0;
+  const ids = new Map<string, string>();
+  for (const block of doc.blocks) {
+    let id = (blockCounter++).toString(16);
+    id = `B${"00000000".slice(id.length) + id}`;
+    ids.set(block.id, id);
+  }
+  for (const mark of doc.marks) {
+    let id = (markCounter++).toString(16);
+    id = `M${"00000000".slice(id.length) + id}`;
+    ids.set(mark.id, id);
+  }
+  for (const block of doc.blocks) {
+    const id = ids.get(block.id);
+    if (id) {
+      block.id = id;
+    }
+    block.attributes = attributesWithStableIds(block.attributes, ids);
+  }
+  for (const mark of doc.marks) {
+    const id = ids.get(mark.id);
+    if (id) {
+      mark.id = id;
+    }
+    mark.attributes = attributesWithStableIds(mark.attributes, ids);
+  }
+
+  return doc;
+}
+
+function unsafe_parseRange(range: Range) {
```

Comment:
In what sense is this unsafe, that start or end might be `NaN`? Would it be better to throw in that case so that it might be handled in a try/catch?